### DISCUSSION TITLE
HRF Estimation with Behavioral Responses (Button Press)?

### DISCUSSION THREAD

**Discussion by alexanderdfung** [cvnlab/GLMsingle#116] [General]
I'm attempting to use GLMsingle for a language task with behavioral responses on some trials. Applying the default GLMsingle pipeline (including HRF estimation) gives me negative (unreasonable) betas in language-selective regions and some relatively late HRFs. Turning off HRF estimation, I get positive betas in language-selective regions. Could the behavioral component be confounding the HRF estimation (are there other known examples of this)?

**Comment by kendrickkay** [comment]
Hi Alex, I would need more details to give a better assessment of this. (For example, how exactly are you incorporating the behavioral responses (in the design matrix)?)  (If you have the folder of figure outputs handy, that could shed some insight. And/or can we see your design matrices?) Also, if you think HRF estimation stuff is of interest, we'd have to look into that more specifically.  One approach is to perform some FIR modeling to see more directly what's going on in the regions you care about.

**Reply by alexanderdfung** [reply]
Hmm... all of the stimuli were sentences presented visually for 2s followed by 4s fixation. Some of the sentences are questions that require a behavioral response. So some of the conditions in the design matrix are associated with a behavioral response while others aren't.

Here is an example model A (ON-OFF) output for one subject (I am plotting the `betasmd` attribute):
![masked_neither_surf_infl_model_A](https://github.com/cvnlab/GLMsingle/assets/46287580/41225f51-faf5-45c2-b9ac-d6bbf48abf45)

Here is a model B (FITHRF) output with HRF fitting (`wantlibrary=1`):
![masked_neither_surf_infl_model_B](https://github.com/cvnlab/GLMsingle/assets/46287580/eb37cd04-e199-4c8d-860f-66df6a7483e9)

Here is a model B output without HRF fitting:
![masked_neither_surf_infl_model_B](https://github.com/cvnlab/GLMsingle/assets/46287580/f0d8ef58-8ae1-452c-a15d-84a0fedd20b2)

Here are the estimated HRFs:
![model_B](https://github.com/cvnlab/GLMsingle/assets/46287580/e96b2b59-19dc-4c07-8ba7-15fb7ef114e4)

Does this give more insight?

**Reply by kendrickkay** [reply]
A few comments.

The first figure looks reasonable, I think.

I agree the second figure looks weird -- but what exactly are you plotting? Are you plotting the mean across all single-trial responses as saved in betasmd?

The third figure looks more similar to first figure, but it is still noticeably different and this is confusing. I would have expected that the first figure and the third figure to be nearly visually identical. (Since the first figure, if i understand correctly, is the %BOLD associated with a simple ON OFF model, and the third figure is a model with a canonical HRF and with a separate regressor for every trial, followed by a simple average across trials.)

The last figure could be informative, but for some reason the R2 is just white (is there a bug?). It would be good to isolate a slice that has clear signals, so we can better interpret what's going on.  Also, the color range for 'average GLM betas' could be tighter to more clearly accentuate the values in voxels with some reasonable signal.

**Comment by iancharest** [comment]
Also please note that we just posted a new release of the Python version of GLMsingle with several bug fixes. Perhaps worth updating your GLMsingle repo to that new release too.
